src/api/ApiService.js

```diff
@@ -0,0 +1,302 @@
+        await setDoc(message, messageData);
+        return message.id;
+    }
+
+    /**
+     * @param {string} chatRoomID
+     * @param {string} messageID
+     */
+    async deleteChatRoomMessage(chatRoomID, messageID) {
+        const message = doc(this.chat_room_messages(chatRoomID), messageID);
+        return deleteDoc(message);
+    }
+
+    /**
+     * @param {string} chatRoomID
+     * @param {import('./types').ChatRoom} chatRoomData
+     */
+    async createChatRoom(chatRoomID, chatRoomData) {
+        const chat_room = doc(this.chat_rooms, chatRoomID);
+        await setDoc(chat_room, chatRoomData);
+        return chat_room.id;
+    }
+
+    /**
+     * @param {string} chatRoomID
+     */
+    async deleteChatRoom(chatRoomID) {
+        const chat_room = doc(this.chat_rooms, chatRoomID);
+        return deleteDoc(chat_room);
+    }
+
+    /**
+     * @param {import('./types').Post} postData
+     */
+    async createPost(postData) {
+        const post = doc(this.posts);
+        await setDoc(post, postData);
+        return post.id;
+    }
+
+    /**
+     * @param {string} postID
+     */
+    async deletePost(postID) {
+        const post = doc(this.posts, postID);
+        return deleteDoc(post);
+    }
+
+    /**
+     * @param {string} postID
+     * @param {import('./types').Post} alreadyFetchedData
+     */
+    async getPost(postID, alreadyFetchedData) {
+        const post = doc(this.posts, postID);
+
+        /**
+         * @type {import('./types').Post}
+         */
+        const data = alreadyFetchedData || (await getDoc(post)).data();
+
+        /**
+         * @type {import('./types').PostFull}
+         */
+        const full_post = {
+            ...data,
+            id: postID,
+            likes: Object.values(data.reactions).reduce(
+                (c, item) => (item === 'LIKE' ? c + 1 : c),
+                0
+            ),
+            dislikes: Object.values(data.reactions).reduce(
+                (c, item) => (item === 'DISLIKE' ? c + 1 : c),
+                0
+            )
+        };
+
+        return full_post;
+    }
+
+    /**
+     * @param {string} postID
+     * @returns {Promise<import('./types').Comment[]>}
+     */
+    async getPostComments(postID) {
+        return (await getDocs(this.post_comments(postID))).docs.map((d) =>
+            d.data()
+        );
+    }
+
+    /**
+     * @callback getPostCommentsCallback
+     * @param {import('./types').Comment[]} comments
+     */
+    /**
+     * @param {string} postID
+     * @param {getPostCommentsCallback} cb
+     */
+    trackGetPostComments(postID, cb) {
+        const q = this.chat_room_messages(postID);
+        return onSnapshot(q, async (d) =>
+            cb(await Promise.all(d.docs.map(async (e) => e.data())))
+        );
+    }
+
+    /**
+     * @param {string} postID
+     * @param {import('./types').Comment} commentData
+     */
+    async createPostComment(postID, commentData) {
```

Comment:
Don't we need a method for editing the post?

Reply:
yup forgot that, thanks